### PR TITLE
Repair specific hosts or specific datacenters

### PR DESCRIPTION
**Pull request by adejanovski**
This PR contains the changes in the server and in spreaper to handle running repairs on specific nodes (to handle repairing nodes after an outage) or specific datacenters.
Both are mutually exclusive.

Repairing specific nodes is done by picking only the segments for which those nodes are replica (not only their primary range).
Repairing specific datacenters is done by specifying the datacenters to the jmx method that triggers repairs. Only nodes within the chosen datacenters will be used as coordinators, otherwise Cassandra would reject the repairs.

An other PR will come soon to bring the associated UI changes.

Please do not merge until the UI PR is merged in this branch.

### CODE REVIEW

**Comment by michaelsembwever** [comment]
Looks like this is broken on Cassandra-2.1
 - https://travis-ci.org/thelastpickle/cassandra-reaper/builds/272200755
 - https://travis-ci.org/thelastpickle/cassandra-reaper/builds/272198278

**Comment by adejanovski** [comment]
Right @michaelsembwever, just uploaded a fix. 
The way the datacenters list is handled in the repair methods has evolved between 2.1 and 2.2 (one wants null, the other one crashes with null 😄  )

**Comment by michaelsembwever** [comment]
@adejanovski do we have a documented style of order of imports? a lot of the files have changes related to re-ordering the import statements.

In lei of any documented style we have, what's the popular style that implicitly exists in the codebase? And are the changes in this PR honouring that or going against it? (it's seems a bit of a mix, but so it's unclear what the tribal standard was here)

I would definitely say though that static imports always come last, and wildcarded imports are avoided.

Beyond that, this is a rather large PR that needs a chunk of time to get through… (and any attempts to reduce the patch size will help a lot)

File: src/server/src/main/java/com/spotify/reaper/resources/RepairRunResource.java [hunk 1]
```diff
@@ -117,67 +120,85 @@ public Response addRepairRun(
...
         LOG.error(ex.getMessage(), ex);
         return Response.status(Response.Status.NOT_FOUND).entity(ex.getMessage()).build();
       }
 
-      RepairUnit theRepairUnit =
-          CommonTools.getNewOrExistingRepairUnit(context, cluster, keyspace.get(), tableNames, incrementalRepair);
+      final Set<String> datacentersToRepair;
+      try {
+        datacentersToRepair = CommonTools.getDatacentersToRepairBasedOnParam(context, cluster,
+            datacentersToRepairParam);
```
Comment (michaelsembwever) [inline src/server/src/main/java/com/spotify/reaper/resources/RepairRunResource.java:167]: nit: each argument on new line, or `.getDatacenter(..)` or a newline.

File: src/server/src/main/java/com/spotify/reaper/resources/RepairRunResource.java [hunk 2]
```diff
@@ -117,67 +120,85 @@ public Response addRepairRun(
...
+      } catch (final IllegalArgumentException ex) {
```
Comment (michaelsembwever) [inline src/server/src/main/java/com/spotify/reaper/resources/RepairRunResource.java:168]: nit: i don't think final exceptions are a style/thing in this codebase?

File: src/server/src/main/java/com/spotify/reaper/resources/RepairRunResource.java [hunk 3]
```diff
@@ -117,67 +120,85 @@ public Response addRepairRun(
...
+        LOG.error(ex.getMessage(), ex);
+        return Response.status(Response.Status.NOT_FOUND).entity(ex.getMessage()).build();
+      }
+
+      final RepairUnit theRepairUnit =
+          CommonTools.getNewOrExistingRepairUnit(context, cluster, keyspace.get(), tableNames, incrementalRepair,
```
Comment (michaelsembwever) [inline src/server/src/main/java/com/spotify/reaper/resources/RepairRunResource.java:174]: nit: each argument on newline, or `.getNew…`

File: src/server/src/main/java/com/spotify/reaper/resources/RepairRunResource.java [hunk 4]
```diff
@@ -265,50 +295,50 @@ public Response modifyRunState(
...
-    for(RepairRun run:repairRuns){
-    	if(!run.getId().equals(repairRunId) && run.getRunState().equals(RunState.RUNNING)){
-    		String errMsg = "repair unit already has run " + run.getId() + " in RUNNING state";
-		    LOG.error(errMsg);
-		    return Response.status(Response.Status.CONFLICT).entity(errMsg).build();
-    	}
+    final Collection<RepairRun> repairRuns = context.storage.getRepairRunsForUnit(repairRun.get().getRepairUnitId());
+
+    for(final RepairRun run:repairRuns){
+      if(!run.getId().equals(repairRunId) && run.getRunState().equals(RunState.RUNNING)){
```
Comment (michaelsembwever) [inline src/server/src/main/java/com/spotify/reaper/resources/RepairRunResource.java:317]: nit: whitespace required in two lines above

File: src/server/src/main/java/com/spotify/reaper/resources/RepairRunResource.java [hunk 5]
```diff
@@ -520,54 +550,54 @@ public Set splitStateParam(Optional<String> state) {
...
+      }
+      // Need to get the RepairUnit before it's possibly deleted.
+      final Optional<RepairUnit> unitPossiblyDeleted =
+          context.storage.getRepairUnit(runToDelete.get().getRepairUnitId());
+      final int segmentsRepaired =
+          context.storage.getSegmentAmountForRepairRunWithState(runId, RepairSegment.State.DONE);
+      final Optional<RepairRun> deletedRun = context.storage.deleteRepairRun(runId);
+      if (deletedRun.isPresent()) {
+        final RepairRunStatus repairRunStatus =
+            new RepairRunStatus(deletedRun.get(), unitPossiblyDeleted.get(), segmentsRepaired);
```
Comment (michaelsembwever) [inline src/server/src/main/java/com/spotify/reaper/resources/RepairRunResource.java:584]: nit: this statement, and those on 579 and 577, can wrap with the `=` starting the new line. ref https://github.com/thelastpickle/cassandra-reaper/commit/99aacd057693dee9967bf056e2dbf37d48a03930#commitcomment-23993308

File: src/server/src/main/java/com/spotify/reaper/resources/view/RepairRunStatus.java [hunk 6]
```diff
@@ -162,7 +170,7 @@ public RepairRunStatus(RepairRun repairRun, RepairUnit repairUnit, int segmentsR
         repairRun.getPauseTime(),
         repairRun.getIntensity(),
         repairUnit.getIncrementalRepair(),
-        repairRun.getRepairParallelism()
+        repairRun.getRepairParallelism(), repairUnit.getNodes(), repairUnit.getDatacenters()
```
Comment (michaelsembwever) [inline src/server/src/main/java/com/spotify/reaper/resources/view/RepairRunStatus.java:173]: nit: new arguments go on new lines

File: src/server/src/main/java/com/spotify/reaper/resources/view/RepairScheduleStatus.java [hunk 7]
```diff
@@ -112,7 +120,7 @@ public RepairScheduleStatus(RepairSchedule repairSchedule, RepairUnit repairUnit
         repairUnit.getIncrementalRepair(),
         repairSchedule.getSegmentCount(),
         repairSchedule.getRepairParallelism(),
-        repairSchedule.getDaysBetween()
+        repairSchedule.getDaysBetween(), repairUnit.getNodes(), repairUnit.getDatacenters()
```
Comment (michaelsembwever) [inline src/server/src/main/java/com/spotify/reaper/resources/view/RepairScheduleStatus.java:123]: nit: new arguments go on new lines

File: src/server/src/main/java/com/spotify/reaper/storage/CassandraStorage.java [hunk 8]
```diff
@@ -405,7 +413,11 @@ public RepairUnit addRepairUnit(RepairUnit.Builder newRepairUnit) {
       if(repairUnitRow.getString("cluster_name").equals(cluster)
           && repairUnitRow.getString("keyspace_name").equals(keyspace)
           && repairUnitRow.getSet("column_families", String.class).equals(columnFamilyNames)){
-        repairUnit = new RepairUnit.Builder(repairUnitRow.getString("cluster_name"), repairUnitRow.getString("keyspace_name"), repairUnitRow.getSet("column_families", String.class), repairUnitRow.getBool("incremental_repair")).build(repairUnitRow.getUUID("id"));
+        repairUnit = new RepairUnit.Builder(repairUnitRow.getString("cluster_name"),
+            repairUnitRow.getString("keyspace_name"), repairUnitRow.getSet("column_families", String.class),
+            repairUnitRow.getBool("incremental_repair"), repairUnitRow.getSet("nodes", String.class),
+            repairUnitRow.getSet("datacenters", String.class))
+                .build(repairUnitRow.getUUID("id"));
```
Comment (michaelsembwever) [inline src/server/src/main/java/com/spotify/reaper/storage/CassandraStorage.java:446]: nit: here, and lines 399 and 387: wrap one argument per line

File: src/server/src/test/java/com/spotify/reaper/resources/CommonToolsTest.java [hunk 9]
```diff
@@ -17,14 +33,70 @@ public void testDateTimeToISO8601() {
...
+    rangeToEndpoint.put(Arrays.asList("2", "4"), Arrays.asList("node1", "node2", "node3"));
+    rangeToEndpoint.put(Arrays.asList("4", "6"), Arrays.asList("node1"));
+    rangeToEndpoint.put(Arrays.asList("6", "8"), Arrays.asList("node1", "node2"));
+
+    Map<String, List<RingRange>> endpointToRangeMap = CommonTools.buildEndpointToRangeMap(rangeToEndpoint);
+
+    assertEquals(endpointToRangeMap.entrySet().size(), 3);
+    assertEquals(endpointToRangeMap.get("node1").size(), 4);
+    assertEquals(endpointToRangeMap.get("node2").size(), 3);
+    assertEquals(endpointToRangeMap.get("node3").size(), 2);
```
Comment (michaelsembwever) [inline src/server/src/test/java/com/spotify/reaper/resources/CommonToolsTest.java:59]: with `Assert.assertEquals(..)` the expected value is the first parameter.

this is one of the reasonings to use the more elegant and fluid Assert4j library and its `Assertions` class, which is already in the classpath and used in places.

**Comment by michaelsembwever** [comment]
nit-picks at a higher/broad level are:

 - procedural growth of codebase: it seems we're adding new features and functionality via continuously adding more parameters to methods. i think we need to be thinking more about the design of the code base and how to use more IoC to avoid this procedural programming style (that's only going to complicate and brittle the code),
 - lack of IoC around handling different Cassandra versions. `if` conditionals i'm not happy about.
 - method parameters are being "train-wrecked" wrapped. i'd rather see each parameter (including the first!) placed on a new line, when line wrapping is required. reference: https://github.com/thelastpickle/cassandra-reaper/commit/99aacd057693dee9967bf056e2dbf37d48a03930#commitcomment-23993308

none of these are blocking the PR though. for example the lack of IoC is a broader issue outside of this specific PR, but it worries me the creep we are seeing. the number of parameters in a lot of the methods are an issue (imho) and we should have a defined approach in place and be leaving the campfire a little cleaner each time we touch it.

i've added a number of nits comments through the code as well, these arn't conclusive (there's many more occurrences in the patch of the things i've nitted about), and most of them are trivial and non-blocking (but to address them would be much appreciated).

otherwise from what i've seen in the patch it looks ok.
i am curious about…

 - what was the outcome with dealing with segment counts and how to calculate that when specific hosts/data-centers are specified?
 - how does this functionality work, and differ, across the different `datacenterAvailability` modes?

**Review by michaelsembwever — APPROVED** [review]

**Pull request merged**